### ISSUE TITLE
Investigate OutOfMemory Exception happening woth Regex.Match for some patterns in the interpreted engine.

### ISSUE THREAD

**Issue by joperezr** [dotnet/runtime#63385]
There are some Regex pattern and input pairs that can trigger an OOM Exception with the Interpreted engine only. The issue seems to be that these crafted inputs are causing the interpreted engine to fail to exit out from a loop and increment the size of runstack array on RegexRunner until eventually an OOM happens. This doesn’t repro on the Compiled, NonBacktracking or source generated engines. 

One example pattern that can be problematic is `(|)+?$` which can repro with pretty much all non empty inputs, for example `11`.

This issue is to track the investigation and resolution of this issue.

**Comment by joperezr** [comment]
I just tried this out in preview 7 and I can't seem to be able to reproduce the issue any longer, so I'll go ahead close this for now.

**Comment by stephentoub** [comment]
> I just tried this out in preview 7 and I can't seem to be able to reproduce the issue any longer, so I'll go ahead close this for now.

I can't repro it even on .NET 6.  Are we sure the cited pattern / input are the right ones?

**Comment by stephentoub** [comment]
I believe this is still an issue.  Try:
```C#
using System;
using System.Text.RegularExpressions;

var r = new Regex(@"(^+?)?()");
r.Match("1");
```

**Comment by joperezr** [comment]
You are right, that pattern does repro in both 7.0 and 6.0. Given this is not a regression from 6 and the limited time we have for 7.0, I'm adjusting the milestone.
